routes/s3handler.js

```diff
@@ -0,0 +1,239 @@
+            parsedMinSize = condition[1];
+            parsedMaxSize = condition[2];
+        }
+    });
+
+    isValid = bucket === expectedBucket;
+
+    // If expectedMinSize and expectedMax size are not null, then
+    // ensure that the client and server have agreed upon the exact same
+    // values.
+    if (expectedMinSize != null && expectedMaxSize != null) {
+        isValid = isValid && (parsedMinSize === expectedMinSize.toString())
+            && (parsedMaxSize === expectedMaxSize.toString());
+    }
+
+    return isValid;
+}
+
+// After the file is in S3, make sure it isn't too big.
+// Omit if you don't have a max file size, or add more logic as required.
+function verifyFileInS3(req, res) {
+    function headReceived(err, data) {
+        if (err) {
+            res.status(500);
+            console.log(err);
+            res.end(JSON.stringify({error: "Problem querying S3!"}));
+        }
+        else if (expectedMaxSize != null && data.ContentLength > expectedMaxSize) {
+            res.status(400);
+            res.write(JSON.stringify({error: "Too big!"}));
+            deleteFile(req.body.bucket, req.body.key, function(err) {
+                if (err) {
+                    console.log("Couldn't delete invalid file!");
+                }
+
+                res.end();
+            });
+        }
+        else {
+            res.end();
+        }
+    }
+
+    callS3("head", {
+        bucket: req.body.bucket,
+        key: req.body.key
+    }, headReceived);
+}
+
+function getV2SignatureKey(key, stringToSign) {
+    var words = CryptoJS.HmacSHA1(stringToSign, key);
+    return CryptoJS.enc.Base64.stringify(words);
+}
+
+function getV4SignatureKey(key, dateStamp, regionName, serviceName, stringToSign) {
+    var kDate = CryptoJS.HmacSHA256(dateStamp, "AWS4" + key),
+        kRegion = CryptoJS.HmacSHA256(regionName, kDate),
+        kService = CryptoJS.HmacSHA256(serviceName, kRegion),
+        kSigning = CryptoJS.HmacSHA256("aws4_request", kService);
+
+    return CryptoJS.HmacSHA256(stringToSign, kSigning).toString();
+}
+
+function deleteFile(bucket, key, callback) {
+    callS3("delete", {
+        bucket: bucket,
+        key: key
+    }, callback);
+}
+
+function callS3(type, spec, callback) {
+    s3[type + "Object"]({
+        Bucket: spec.bucket,
+        Key: spec.key
+    }, callback)
+}
+
+router.use(bodyParser.json()); // support json encoded bodies
+router.use(bodyParser.urlencoded({ extended: true })); // support encoded bodies
+
+router.post("/", function(req, res) {
+    if (typeof req.query.success !== "undefined") {
+        verifyFileInS3(req, res);
+    }
+    else {
+        signRequest(req, res);
+    }
+});
+
+router.post("/success", async function(req, res) {
+  console.log('s3 upload success!')
+  // console.log('req:', req.body)
+  const { bucket, key } = req.body
+
+  //TODO: Insert into the db once schema is finalized?
```

Comment:
Maybe make issues for these on GitHub to keep track since it's hard to remember they're in the code.